src/contrib/cluster/Akka.Cluster.Tools.Tests.MultiNode/ClusterClient/ClusterClientDiscoverySpec.cs

```diff
@@ -0,0 +1,297 @@
+        {
+            Join(_config.Second, _config.Second);
+            RunOn(() =>
+            {
+                var service = Sys.ActorOf(EchoActor.Props(this), "testService");
+                ClusterClientReceptionist.Get(Sys).RegisterService(service);
+                AwaitMembersUp(1);
+            }, _config.Second);
+            
+            EnterBarrier("cluster-restarted");
+            
+            RunOn(() =>
+            {
+                var address = GetAddress(_config.Second);
+                _discoveryService.TryAddEndpoint("test-cluster", new ServiceDiscovery.ResolvedTarget(address.Host, address.Port));
+                
+                var resolved = _discoveryService.Lookup(new Lookup("test-cluster"), TimeSpan.FromSeconds(1)).Result;
+                resolved.Addresses.Count.Should().Be(1);
+            }, _config.Client);
+            EnterBarrier("discovery-entry-updated");
+        }
+
+        private void ClusterClient_must_re_establish_on_cluster_restart()
+        {
+            RunOn(() =>
+            {
+                Within(TimeSpan.FromSeconds(5), () =>
+                {
+                    AwaitAssert(() =>
+                    {
+                        _clusterClient.Tell(GetContactPoints.Instance, TestActor);
+                        var contacts = ExpectMsg<ContactPoints>(TimeSpan.FromSeconds(1)).ContactPointsList;
+                        contacts.Count.Should().Be(1);
+                        contacts.First().Address.Should().Be(Node(_config.Second).Address);
+                    }, RemainingOrDefault);
+                });
+
+                _clusterClient.Tell(new ClusterClient.Send("/user/testService", "hello", localAffinity: true));
+                ExpectMsg<string>().Should().Be("hello");
+               
+            }, _config.Client);
+            EnterBarrier("re-establish-successful");
+        }
+
+        private void ClusterClient_must_simulate_a_cluster_forced_shutdown()
+        {
+            RunOn(() =>
+            {
+                // simulate a hard shutdown
+                TestConductor.Exit(_config.Second, 0).Wait();
+            }, _config.Client);
+            EnterBarrier("hard-shutdown-and-discovery-entry-updated");
+        }
+
+        private void ClusterClient_third_node_formed_a_cluster()
+        {
+            Join(_config.Third, _config.Third);
+            RunOn(() =>
+            {
+                var service = Sys.ActorOf(EchoActor.Props(this), "testService");
+                ClusterClientReceptionist.Get(Sys).RegisterService(service);
+                AwaitMembersUp(1);
+            }, _config.Third);
+            
+            EnterBarrier("cluster-restarted");
+            
+            RunOn(() =>
+            {
+                var address = GetAddress(_config.Third);
+                _discoveryService.TryAddEndpoint("test-cluster", new ServiceDiscovery.ResolvedTarget(address.Host, address.Port));
```

Comment:
It might be more interesting if we had a 2 node cluster (have the second node join AFTER the `ClusterClient` has been able to contact the first node successfully) and see if it fails over automatically.

Reply:
Nevermind, that's all handled by the `ClusterClient` - which only kills itself when there are zero available